File: website/alg.js
function findRecipes(items, rating, ingredientMap, countMap){
    const user_ing = items
    const user_rat = rating / 100

    const Ing_Rec = ingredientMap; // Maps ingredients to a list of recipes it is included in
    const Rec_Tot = countMap; // Maps recipes to the number of ingredients required
    const Rec_Num = {}; // Maps recipes to the number of ingredients the user has

    for (const ingredient of user_ing) {
        for (const recipe of Ing_Rec[ingredient]) {
            if (!Rec_Num[recipe]) {
                Rec_Num[recipe] = 1;
            } else {
                Rec_Num[recipe]++;
            }
        }
    }

    const Rat_Rec = {}; // Maps the rating of the recipe to recipes

    for (const recipe in Rec_Num) {
        const rating = Rec_Num[recipe] / Rec_Tot[recipe];

        if (!Rat_Rec[rating]) {
            Rat_Rec[rating] = [];
        }
        Rat_Rec[rating].push(recipe);
    }

    const keys = Object.keys(Rat_Rec).sort((a, b) => b - a);

    /*for (const key of keys) {
        console.log(key, Rat_Rec[key]);
    }*/

    const output = []
    for (const key of keys){
        if (key < user_rat)
            continue
        for (const recipe of Rat_Rec[key]){
            output.push(recipe)
        }
    }

    console.log(output)
}

File: algorithm/alg.py
import csv

Ing_Rec = {} # Maps ingredients to list of recipes it is included in
Rec_Tot = {} # Maps recipes to the number of ingredients required

# Read the CSV file
with open('../Datasets/Testing/genRecipes.csv', newline='', encoding='utf-8') as csvfile:
    reader = csv.DictReader(csvfile)

    # Print only the Recipe names
    for row in reader:
        count = 0
        for item in row['ingredients'].split(', '):
            ingredient = item.replace('[', '').replace(']', '').replace('\'', '')
           
            if ingredient not in Ing_Rec:
                Ing_Rec[ingredient] = list()
            Ing_Rec[ingredient].append(row['title'])
            count += 1
        Rec_Tot[row['title']] = count


user_ing = ['Pasta', 'Ground Beef', 'Tomato Sauce', 'Onion', 'Garlic', 'Olive Oil', 'Salt', 'Pepper', 'Cabbage Leaves'] # Testing list
Rec_Num = {} # Maps recipes to the number of ingredients the user has

for ingredient in user_ing:
    for recipe in Ing_Rec[ingredient]:
        if recipe not in Rec_Num:
            Rec_Num[recipe] = 1
        else:
            Rec_Num[recipe] += 1


Rat_Rec = {} # Maps rating of recipe to recipes

for recipe in Rec_Num.keys():
    rating = Rec_Num[recipe] / Rec_Tot[recipe]
    if rating not in Rat_Rec:
        Rat_Rec[rating] = list()
    Rat_Rec[rating].append(recipe)


keys = sorted(Rat_Rec.keys(), reverse=True)

for key in keys:
    print(key, Rat_Rec[key])

